Modules/posixmodule.c

```diff
@@ -10928,28 +10950,32 @@ os_sendfile_impl(PyObject *module, int out_fd, int in_fd, PyObject *offobj,
     } while (ret < 0 && errno == EINTR && !(async_err = PyErr_CheckSignals()));
     _Py_END_SUPPRESS_IPH
 
-    if (sf.headers != NULL)
-        iov_cleanup(sf.headers, hbuf, sf.hdr_cnt);
-    if (sf.trailers != NULL)
-        iov_cleanup(sf.trailers, tbuf, sf.trl_cnt);
-
     if (ret < 0) {
         if ((errno == EAGAIN) || (errno == EBUSY)) {
             if (sbytes != 0) {
                 // some data has been sent
                 goto done;
             }
-            else {
-                // no data has been sent; upper application is supposed
-                // to retry on EAGAIN or EBUSY
-                return posix_error();
-            }
+            // no data has been sent; upper application is supposed
+            // to retry on EAGAIN or EBUSY
         }
-        return (!async_err) ? posix_error() : NULL;
```

Comment:
I would prefer to use return on error. The new code is harder to think about, the control flow is less usual.

Reply:
Done.